src/renderer/tvgTrimPath.cpp

```diff
@@ -0,0 +1,337 @@
+            case PathCommand::Close:
+                return length(pts - 1, &moveTo);
+        }
+        return 0.0f;
+    };
+
+    auto _shift = [&]() -> void {
+        switch (*cmds) {
+            case PathCommand::MoveTo:
+                moveTo = *pts;
+                moveToTrimmed = *pts;
+                ++pts;
+                break;
+            case PathCommand::LineTo:
+                ++pts;
+                break;
+            case PathCommand::CubicTo:
+                pts += 3;
+                break;
+            case PathCommand::Close:
+                break;
+        }
+        ++cmds;
+    };
+
+    bool start = true;
+    for (uint32_t i = 0; i < inCmdsCnt; ++i) {
+        auto dLen = _length();
+        //skip the segment
+        if (len + dLen <= trimStart || len > trimEnd) {}
+        else if (len <= trimStart) {
+            //cut the segment at the beginning and at the end
+            if (len + dLen > trimEnd) _trimAt(cmds, pts, moveToTrimmed, trimStart - len, trimEnd - trimStart, start, outCmds, outPts);
+            //cut the segment at the beginning
+            else _trimAt(cmds, pts, moveToTrimmed, trimStart - len, len + dLen - trimStart, start, outCmds, outPts);
+            start = false;
+        //cut the segment at the end
+        } else if (len + dLen > trimEnd) {
+            _trimAt(cmds, pts, moveTo, 0.0f, trimEnd - len, start, outCmds, outPts);
+            start = true;
+        //add the whole segment
+        } else _add(cmds, pts, moveTo, start, outCmds, outPts);
+
+        len += dLen;
+        _shift();
+    }
+}
+
+
+static void _trim(const PathCommand* inCmds, uint32_t inCmdsCnt, const Point* inPts, uint32_t inPtsCnt, float begin, float end, Array<PathCommand>& outCmds, Array<Point>& outPts)
+{
+    auto totalLength = _pathLength(inCmds, inCmdsCnt, inPts, inPtsCnt);
+    auto trimStart = begin * totalLength;
+    auto trimEnd = end * totalLength;
+
+    if (trimStart > trimEnd) {
+        _trimPath(inCmds, inCmdsCnt, inPts, inPtsCnt, trimStart, totalLength, outCmds, outPts);
+        _trimPath(inCmds, inCmdsCnt, inPts, inPtsCnt, 0.0f, trimEnd, outCmds, outPts);
+    } else {
+        _trimPath(inCmds, inCmdsCnt, inPts, inPtsCnt, trimStart, trimEnd, outCmds, outPts);
+    }
+}
+
+
+/************************************************************************/
+/* External Class Implementation                                        */
+/************************************************************************/
+
+
+bool TrimPath::valid() const
+{
+    if (begin == 0.0f && end == 1.0f) return false;
+    if (fabsf(end - begin) >= 1.0f) return false;
+    return true;
+}
+
+
+bool TrimPath::get(float& begin, float& end) const
+{
+    begin = this->begin;
+    end = this->end;
+
+    if (fabsf(end - begin) >= 1.0f) {
+        begin = 0.0f;
+        end = 1.0f;
+        return false;
+    }
+
+    auto loop = true;
+
+    if (begin > 1.0f && end > 1.0f) loop = false;
```

Comment:
Please double-check if out-of-range trimming exceptions are really necessary here.

Reply:
yes, because in lottie there is offset value (and in api we let for all possible values of begin and end)

[sample.json](https://github.com/user-attachments/files/18679959/sample.json)
tvg vs lottie-web:
![tvg_trim](https://github.com/user-attachments/assets/4d8413c9-9001-49a5-b829-fd9a1cf31d5e)
![lottie_web_trim](https://github.com/user-attachments/assets/5cf82c70-85c0-4ef8-a6f7-8a911a40c97a)

in the file:
- begin: 90%
- end: 0%-100%
- offset: 180%
so as the result during this animation `begin` changes in the 0.5-1.4 range and `end` changes in the 1.4-1.5 range